src/metabase/api/permission_graph.clj

```diff
@@ -7,105 +7,110 @@
 
-(s/def ::schema-name (s/or :kw->str keyword?))
-
-;; {:groups {1 {:data {:schemas {"PUBLIC" ::schema-perms-granular}}}}} =>
-;; {:groups {1 {:data {:schemas {"PUBLIC" {1 :all}}}}}}
-(s/def ::read (s/or :str->kw #{"all" "none"}))
-(s/def ::query (s/or :str->kw #{"all" "none" "segmented"}))
-
-(s/def ::table-perms-granular (s/keys :opt-un [::read ::query]))
-
-(s/def ::table-perms (s/or :str->kw #{"all" "segmented" "none" "full" "limited"}
-                           :identity ::table-perms-granular))
-
-(s/def ::table-graph (s/map-of ::id ::table-perms
-                               :conform-keys true))
-
-(s/def ::schema-perms (s/or :str->kw #{"all" "segmented" "none" "full" "limited"}
-                            :identity ::table-graph))
-
-;; {:groups {1 {:data {:schemas {"PUBLIC" ::schema-perms}}}}}
-(s/def ::schema-graph (s/map-of ::schema-name ::schema-perms
-                                :conform-keys true))
-
-;; {:groups {1 {:data {:schemas ::schemas}}}}
-(s/def ::schemas (s/or :str->kw   #{"all" "segmented" "none" "block" "full" "limited"}
-                       :nil->none nil?
-                       :identity  ::schema-graph))
 
-(s/def ::data (s/keys :opt-un [::native ::schemas]))
-
-(s/def ::download (s/keys :opt-un [::native ::schemas]))
-
-(s/def ::data-model (s/keys :opt-un [::native ::schemas]))
-
-;; We use "yes" and "no" instead of booleans for consistency with the application perms graph, and consistency with the
-;; language used on the frontend.
-(s/def ::details (s/or :str->kw #{"yes" "no"}))
-
-(s/def ::db-perms (s/keys :opt-un [::data ::download ::data-model ::details ::execute]))
-
-(s/def ::db-graph
-  (s/map-of ::id
-            ::db-perms
-            :conform-keys true))
-
-(s/def :metabase.api.permission-graph.data/groups
-  (s/map-of ::id ::db-graph
-            :conform-keys true))
-
-(s/def ::data-permissions-graph
-  (s/keys :req-un [:metabase.api.permission-graph.data/groups]))
+(def ^:private table-perms
+  [:or
+   [:enum :all :segmented :none :full :limited]
+   [:map
+    [:read [:enum :all :none]]
+    [:query [:enum :all :none :segmented]]]])
+
+(def ^:private schema-perms
+  [:or
+   [:keyword {:title "schema name"}]
+   [:map-of id table-perms]])
+
+(def ^:private schema-graph
+  [:map-of
+   [:string {:decode/perm-graph name}]
+   schema-perms])
+
+(def ^:private schemas
+  [:or
+   [:enum :all :segmented :none :block :full :limited]
+   schema-graph])
+
+(def ^:private data-perms
+  [:map
+   [:native {:optional true} native]
+   [:schemas {:optional true} schemas]])
+
+(def ^:private strict-data-perms
+  [:and
+   data-perms
+   [:fn {:error/fn (constantly
+                    (trs "Invalid DB permissions: If you have write access for native queries, you must have full data access."))}
+    (fn [{:keys [native schemas]}]
+      (not (and (= native :write) schemas (not= schemas :all))))]])
```

Comment:
This piece is from the metabase.models.permissions namespace's definition of strict